staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go

```diff
@@ -177,9 +188,23 @@ type deferredResponseWriter struct {
 	hasWritten bool
 	hw         http.ResponseWriter
 	w          io.Writer
+
+	trace *utiltrace.Trace
 }
 
 func (w *deferredResponseWriter) Write(p []byte) (n int, err error) {
+	if w.trace != nil {
+		// This Step usually wraps in-memory object serialization.
+		w.trace.Step("About to start writing response", utiltrace.Field{"size", len(p)})
+
+		firstWrite := !w.hasWritten
+		defer func() {
+			w.trace.Step("Write call finished",
```

Comment:
How many Write() calls we can have per serialization?

I think protobuf encoder does O(1), but what about json/yaml? I would like to ensure that we can't end up with gazillions of those and we won't introduce visible overhead with those...

Reply:
For json/yaml it's a single call as well (see [1] for yaml, [2] for json) We already use assumption about single call for gzip compression (see comment for defaultGzipThresholdBytes constant)

[1] https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go#L314
[2] https://golang.org/src/encoding/json/stream.go?s=5070:5117#L231 called here: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go#L327